### PR TITLE
[ESIMD] Enable esimd emulator build by default

### PR DESCRIPTION
**Pull request by dongkyunahn-intel**
This PR is to enable ESIMD_EMULATOR build by default.

- PI_APIs for PI info are updated for check-in tests

### CODE REVIEW

File: buildbot/configure.py [hunk 1]
```diff
@@ -55,9 +54,6 @@ def do_configure(args):
     if args.arm:
         llvm_targets_to_build = 'ARM;AArch64'
 
-    if args.enable_esimd_cpu_emulation:
```
Comment (bader) [inline buildbot/configure.py:58]: Please, remove command line option.

**Review by bader — CHANGES_REQUESTED** [review]

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 2]
```diff
@@ -1324,6 +1360,7 @@ pi_result piEnqueueMemBufferReadRect(pi_queue, pi_mem, pi_bool,
 pi_result piEnqueueMemBufferWrite(pi_queue, pi_mem, pi_bool, size_t, size_t,
                                   const void *, pi_uint32, const pi_event *,
                                   pi_event *) {
+  // TODO : intel/llvm_test_suite
```
Comment (kbobrovs) [inline sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:1363]: what does this mean - "// TODO add a test to github.com/intel/llvm-test-suite"?
Please update if so. Even though I don't see much sense in such TODO.

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 3]
```diff
@@ -1623,7 +1674,8 @@ pi_result piextDeviceSelectBinary(pi_device, pi_device_binary *,
 pi_result piextUSMEnqueuePrefetch(pi_queue, const void *, size_t,
                                   pi_usm_migration_flags, pi_uint32,
                                   const pi_event *, pi_event *) {
-  DIE_NO_IMPLEMENTATION;
+  // NOP for prefetch
```
Comment (kbobrovs) [inline sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:1677]: Please explain in the comment why NOP for prefetch is safe.

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 4]
```diff
@@ -528,7 +538,7 @@ pi_result piDeviceGetInfo(pi_device Device, pi_device_info ParamName,
   case PI_DEVICE_INFO_VERSION:
     // CM_EMU release version from
     // https://github.com/intel/cm-cpu-emulation/releases
-    return ReturnValue("1.0.7-CM_EMU");
+    return ReturnValue("1.0");
```
Comment (kbobrovs) [inline sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:541]: this should rather be a query into libcm. at least TODO should be added

File: sycl/source/detail/scheduler/commands.cpp [hunk 5]
```diff
@@ -2157,10 +2157,12 @@ cl_int ExecCGCommand::enqueueImp() {
       } else {
         assert(MQueue->getPlugin().getBackend() ==
                backend::ext_intel_esimd_emulator);
+        // Dims==0 for 'single_task() - void(void) type'
+        uint32_t Dims = (Args.size() > 0) ? NDRDesc.Dims : 0;
```
Comment (kbobrovs) [inline sycl/source/detail/scheduler/commands.cpp:2161]: AFAICT, single_task is not supported. Should this cause DIE_NO_IMPLEMENTATION?

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp (hunk 2 above)
Comment (dongkyunahn-intel) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:1363]: Will do.

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp (hunk 4 above)
Comment (dongkyunahn-intel) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:541]: I'll check if the version number can be retrieved from CM and apply it if possible.

And, please note that the device info version format should be 'X.Y'. It is required from one of tests in intel/llvm-test-suite.

File: sycl/source/detail/scheduler/commands.cpp (hunk 5 above)
Comment (dongkyunahn-intel) [reply sycl/source/detail/scheduler/commands.cpp:2161]: 'single_task()' is widely used in intel/llvm-test-suite. `ESIMD/SYCL/printf.cpp` uses 'single_task()'. I would like to make sure if it is going to be supported or not.

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp (hunk 3 above)
Comment (dongkyunahn-intel) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:1677]: I'll replace it with TODO like above ones.

**Comment by kbobrovs** [comment]
Note on running SYCL kernels:
ESIMD emulator by definition is another kind of "gpu" device, which can run only ESIMD kernels, it can't run normal SYCL kernels. We need a way to detect that ESIMD EMU plugin is asked to run a non-ESIMD kernel (i.e. sycl_explicit_simd attribute is missing for it), and issue a meaningful error. This might require extending the PI kernel invocation interface to propagate such kernel attributes/properties. IIRC, we added special new invocation interface to PI to invoke kernels bypassing the "normal" mechanism, so we could extend it further to propagate needed info. @romanovvlad, @smaslov-intel - do you agree or have other comments?

**Comment by romanovvlad** [comment]
> Note on running SYCL kernels: ESIMD emulator by definition is another kind of "gpu" device, which can run only ESIMD kernels, it can't run normal SYCL kernels. We need a way to detect that ESIMD EMU plugin is asked to run a non-ESIMD kernel (i.e. sycl_explicit_simd attribute is missing for it), and issue a meaningful error. This might require extending the PI kernel invocation interface to propagate such kernel attributes/properties. IIRC, we added special new invocation interface to PI to invoke kernels bypassing the "normal" mechanism, so we could extend it further to propagate needed info. @romanovvlad, @smaslov-intel - do you agree or have other comments?

If we convert `sycl_explicit_simd` attribute to a compile time property(see https://github.com/intel/llvm/pull/4937) and starting report something like piDeviceGetInfo(PI_SUPPORT_ESIMD) == true from plugins that support esimd execution model we can have the wanted diagnostic in the SYCL RT alongside with other checks for kernel requirements vs device capabilities.

**Comment by kbobrovs** [comment]
> > Note on running SYCL kernels: ESIMD emulator by definition is another kind of "gpu" device, which can run only ESIMD kernels, it can't run normal SYCL kernels. We need a way to detect that ESIMD EMU plugin is asked to run a non-ESIMD kernel (i.e. sycl_explicit_simd attribute is missing for it), and issue a meaningful error. This might require extending the PI kernel invocation interface to propagate such kernel attributes/properties. IIRC, we added special new invocation interface to PI to invoke kernels bypassing the "normal" mechanism, so we could extend it further to propagate needed info. @romanovvlad, @smaslov-intel - do you agree or have other comments?
> 
> If we convert `sycl_explicit_simd` attribute to a compile time property(see #4937) and starting report something like piDeviceGetInfo(PI_SUPPORT_ESIMD) == true from plugins that support esimd execution model we can have the wanted diagnostic in the SYCL RT alongside with other checks for kernel requirements vs device capabilities.
> 

I'm not sure if #4937 will work for emulator, as device code does not go through SPIRV or even device compiler.

> BTW.
> 
> > > IIRC, we added special new invocation interface to PI to invoke kernels bypassing the "normal" mechanism
> 
> I believe we have avoided adding new invocations interface, could you please share a link to the line where it's defined if we have it?

Looks like my IIRC is wrong. I see we use piEnqueueKernelLaunch.
Maybe SYCL RT could see if there is isESIMD property on device binary being passed to ESIMD emulator BE and report an error if it is missing.

**Comment by romanovvlad** [comment]
> > > Note on running SYCL kernels: ESIMD emulator by definition is another kind of "gpu" device, which can run only ESIMD kernels, it can't run normal SYCL kernels. We need a way to detect that ESIMD EMU plugin is asked to run a non-ESIMD kernel (i.e. sycl_explicit_simd attribute is missing for it), and issue a meaningful error. This might require extending the PI kernel invocation interface to propagate such kernel attributes/properties. IIRC, we added special new invocation interface to PI to invoke kernels bypassing the "normal" mechanism, so we could extend it further to propagate needed info. @romanovvlad, @smaslov-intel - do you agree or have other comments?
> > 
> > 
> > If we convert `sycl_explicit_simd` attribute to a compile time property(see #4937) and starting report something like piDeviceGetInfo(PI_SUPPORT_ESIMD) == true from plugins that support esimd execution model we can have the wanted diagnostic in the SYCL RT alongside with other checks for kernel requirements vs device capabilities.
> 
> I'm not sure if #4937 will work for emulator, as device code does not go through SPIRV or even device compiler.

With this properties we could convert all attributes to properties, so they are known to the SYCL RT without compiler assistance.

**Comment by kbobrovs** [comment]
> > > > Note on running SYCL kernels: ESIMD emulator by definition is another kind of "gpu" device, which can run only ESIMD kernels, it can't run normal SYCL kernels. We need a way to detect that ESIMD EMU plugin is asked to run a non-ESIMD kernel (i.e. sycl_explicit_simd attribute is missing for it), and issue a meaningful error. This might require extending the PI kernel invocation interface to propagate such kernel attributes/properties. IIRC, we added special new invocation interface to PI to invoke kernels bypassing the "normal" mechanism, so we could extend it further to propagate needed info. @romanovvlad, @smaslov-intel - do you agree or have other comments?
> > > 
> > > 
> > > If we convert `sycl_explicit_simd` attribute to a compile time property(see #4937) and starting report something like piDeviceGetInfo(PI_SUPPORT_ESIMD) == true from plugins that support esimd execution model we can have the wanted diagnostic in the SYCL RT alongside with other checks for kernel requirements vs device capabilities.
> > 
> > 
> > I'm not sure if #4937 will work for emulator, as device code does not go through SPIRV or even device compiler.
> 
> With this properties we could convert all attributes to properties, so they are known to the SYCL RT without compiler assistance.

To get properties we need device binary. Device binary is not used with the ESIMD emulator, I believe. Should we at all require presence of device binary when running on host device (ESIMD emulator shares a lot with it) ?

**Comment by romanovvlad** [comment]
> > > > > Note on running SYCL kernels: ESIMD emulator by definition is another kind of "gpu" device, which can run only ESIMD kernels, it can't run normal SYCL kernels. We need a way to detect that ESIMD EMU plugin is asked to run a non-ESIMD kernel (i.e. sycl_explicit_simd attribute is missing for it), and issue a meaningful error. This might require extending the PI kernel invocation interface to propagate such kernel attributes/properties. IIRC, we added special new invocation interface to PI to invoke kernels bypassing the "normal" mechanism, so we could extend it further to propagate needed info. @romanovvlad, @smaslov-intel - do you agree or have other comments?
> > > > 
> > > > 
> > > > If we convert `sycl_explicit_simd` attribute to a compile time property(see #4937) and starting report something like piDeviceGetInfo(PI_SUPPORT_ESIMD) == true from plugins that support esimd execution model we can have the wanted diagnostic in the SYCL RT alongside with other checks for kernel requirements vs device capabilities.
> > > 
> > > 
> > > I'm not sure if #4937 will work for emulator, as device code does not go through SPIRV or even device compiler.
> > 
> > 
> > With this properties we could convert all attributes to properties, so they are known to the SYCL RT without compiler assistance.
> 
> To get properties we need device binary. Device binary is not used with the ESIMD emulator, I believe. Should we at all require presence of device binary when running on host device (ESIMD emulator shares a lot with it) ?

I'm referring to `sycl::ext::oneapi::property_list` and `sycl::ext::oneapi::property_value`(see https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/PropertyList/SYCL_EXT_ONEAPI_property_list.asciidoc and https://github.com/intel/llvm/pull/4937/files?short_path=8b8203c#diff-8b8203c67c38c950b9c19edd0aa137e67e3683c84fc454fe2982ebb19e7a47b1)

File: sycl/source/detail/scheduler/commands.cpp (hunk 5 above)
Comment (kbobrovs) [reply sycl/source/detail/scheduler/commands.cpp:2176]: But your code implies it is not supported yet. So DIE_NO_IMPLEMENTATION should happen on attempt to use it.

**Comment by kbobrovs** [comment]
> I'm referring to sycl::ext::oneapi::property_list and sycl::ext::oneapi::property_value(see https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/PropertyList/SYCL_EXT_ONEAPI_property_list.asciidoc and https://github.com/intel/llvm/pull/4937/files?short_path=8b8203c#diff-8b8203c67c38c950b9c19edd0aa137e67e3683c84fc454fe2982ebb19e7a47b1)

How these would be propagated to SYCL runtime?

File: sycl/source/detail/scheduler/commands.cpp (hunk 5 above)
Comment (dongkyunahn-intel) [reply sycl/source/detail/scheduler/commands.cpp:2176]: `DIE_NO_IMPLEMENTATION` is defined in `sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp`. Maybe execution attempts for `single_task()` return `PI_INVALID_WORK_DIMENSION` here until it is supported?

**Review by bader — APPROVED** [review]
buildbot/configure.py looks okay, but I have one question though: should we have a switch to disable ESIMD emulator build? I suppose it might be useful for developers not involved in ESIMD development.

File: sycl/source/handler.cpp [hunk 6]
```diff
@@ -206,10 +207,22 @@ event handler::finalize() {
         MHostKernel->call(
             MNDRDesc, (NewEvent) ? NewEvent->getHostProfilingInfo() : nullptr);
         return CL_SUCCESS;
+      } else {
```
Comment (bader) [inline sycl/source/handler.cpp:210]: Please, don't use `else` after return.
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f56-avoid-unnecessary-condition-nesting

Comment (dongkyunahn-intel) [reply sycl/source/handler.cpp]: Updated with single-point-of-return.

**Review by vladimirlaz — APPROVED** [review]

**Comment by bader** [comment]
> buildbot/configure.py looks okay, but I have one question though: should we have a switch to disable ESIMD emulator build? I suppose it might be useful for developers not involved in ESIMD development.

@dongkyunahn-intel, @kbobrovs, what do you think about flipping the switch instead of removing it?

**Comment by dongkyunahn-intel** [comment]
> buildbot/configure.py looks okay, but I have one question though: should we have a switch to disable ESIMD emulator build? I suppose it might be useful for developers not involved in ESIMD development.

Do other backend types have disabling option like you suggested? If not, I don't want to have it only for ESIMD emulator.

**Comment by kbobrovs** [comment]
> > > buildbot/configure.py looks okay, but I have one question though: should we have a switch to disable ESIMD emulator build? I suppose it might be useful for developers not involved in ESIMD development.
> > 
> > 
> > Do other backend types have disabling option like you suggested? If not, I don't want to have it only for ESIMD emulator.
> 
> HIP and CUDA backend builds are controlled by dedicated options.

Makes sense to me too.

**Comment by vladimirlaz** [comment]
I am trying to build esimd_emulator locally and got the following error:
```
llvm/build/tools/sycl/plugins/esimd_emulator/cm-emu-prefix/src/cm-emu/cmake/../common/linux/cm_include.h:22:10: fatal error: va/va.h: No such file or directory
   22 | #include "va/va.h"
      |          ^~~~~~~~~
compilation terminated.
```
Does this build brings new dependency: *libva-dev* package to be installed on ?

**Comment by bader** [comment]
AFAIK, `libva-dev` is not the only dependency. See https://github.com/intel/llvm/pull/5212 for the full list.

**Review by bader — APPROVED** [review]
buildbot/configure.py changes look good to me.
We need someone from @intel/llvm-reviewers-runtime to approve.

**Review by vladimirlaz — APPROVED** [review]

**Comment by alexbatashev** [comment]
@bader working on it

**Pull request merged**